plugins/Oscilloscope/OscilloscopeGraph.cpp

```diff
@@ -0,0 +1,168 @@
+	float phase = m_controls->m_phaseModel.value();
+	int windowSizeFrames = m_ringBuffer.size() * m_controls->m_lengthModel.value() / m_controls->m_lengthModel.maxValue();
+	int framesPerPixel = std::max(1, windowSizeFrames / width());
+
+	int windowStartIndex = m_ringBufferIndex + (m_ringBuffer.size() - windowSizeFrames) + (phase) * m_ringBuffer.size();
+	// Quantize start index to prevent flickering (round up to prevent going below the current ring buffer index)
+	int frameOffsetWithinPixel = windowStartIndex % framesPerPixel;
+	windowStartIndex += framesPerPixel - frameOffsetWithinPixel;
+	// And subtract that amount from the window size to prevent going past off the right
+	windowSizeFrames -= framesPerPixel - frameOffsetWithinPixel;
+
+	p.setPen(m_minorLineColor);
+	p.drawLine(0, height() / 2, width(), height() / 2);
+
+	p.setPen(m_clippingLineColor);
+	p.drawLine(0, height() / 2 * (1 - amp), width(), height() / 2 * (1 - amp));
+	p.drawLine(0, height() / 2 * (1 + amp), width(), height() / 2 * (1 + amp));
+
+	const bool hq = framesPerPixel > 1;
+	const int xoffset = !hq * framesPerPixel;
+	const float xscale = 1.f / (windowSizeFrames - framesPerPixel);
+	const bool stereo = m_controls->m_stereoModel.value();
+
+	auto drawWaveform = [&](QColor& color, auto&& getChannel)
+	{
+		p.setCompositionMode(QPainter::CompositionMode_Plus);
+		p.setPen(color);
+		for (int f = 0; f < windowSizeFrames - framesPerPixel; f += framesPerPixel)
+		{
+			const int currentIndex = (windowStartIndex + f) % m_ringBuffer.size();
+			const int nextIndex = (windowStartIndex + f + framesPerPixel) % m_ringBuffer.size();
+
+			float maxValue = getChannel(m_ringBuffer[hq ? currentIndex : nextIndex]);
+			float minValue = getChannel(m_ringBuffer[currentIndex]);
+
+			for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
+			{
+				maxValue = std::max(maxValue, getChannel(m_ringBuffer[i]));
+				minValue = std::min(minValue, getChannel(m_ringBuffer[i]));
+			}
+
+			p.drawLine(
+				width() * f * xscale,
+				height() * (1 - minValue * amp) / 2,
+				width() * (f + xoffset) * xscale,
+				height() * (1 - maxValue * amp) / 2
+			);
+		}
+	};
+
+	if (!stereo)
+	{
+		drawWaveform(m_monoColor, [](const SampleFrame& f){ return f.average(); });
+	}
+	else
+	{
+		drawWaveform(m_leftColor, [](const SampleFrame& f){ return f.left(); });
+		drawWaveform(m_rightColor, [](const SampleFrame& f){ return f.right(); });
+	}
+}
+
+void OscilloscopeGraph::wheelEvent(QWheelEvent* we)
+{
+	float windowSizeMilliseconds = m_controls->m_lengthModel.value();
+	float phase = m_controls->m_phaseModel.value();
+	float mouseOffset = (1.0f - we->position().x() / width()) * (windowSizeMilliseconds / m_controls->m_lengthModel.maxValue());
+	float zoomAmount = std::clamp(std::exp2(-we->angleDelta().y() / 240.0f), m_controls->m_lengthModel.minValue() / m_controls->m_lengthModel.value(), m_controls->m_lengthModel.maxValue() / m_controls->m_lengthModel.value());
```

Comment:
(Non-blocking)

- `windowSizeMilliseconds`, `mouseOffset`, and `zoomAmount` are never mutated and should be declared `const`.
- `phase` is used exactly once and can be inlined (it's clear that it's the phase even without this variable binding)
- Some of these lines exceed 120 characters in length, breaking them on operators resolves this

```suggestion
	const float windowSizeMilliseconds = m_controls->m_lengthModel.value();
	const float mouseOffset = (1.0f - we->position().x() / width())
		* (windowSizeMilliseconds / m_controls->m_lengthModel.maxValue());
	const float zoomAmount = std::clamp(
		std::exp2(-we->angleDelta().y() / 240.0f),
		m_controls->m_lengthModel.minValue() / m_controls->m_lengthModel.value(),
		m_controls->m_lengthModel.maxValue() / m_controls->m_lengthModel.value()
	);
```